### PR TITLE
fix(lsp): add a document preload file system entry limit

### LINKED ISSUE #18538: `deno repl` hangs with excessive memory usage running from certain CWDs
OS: Windows 10 Version 10.0.19045.2728

From `deno 1.32.1+7722014` and up, running `deno repl` or just `deno` from certain directories causes the deno process to hang with no repl prompt and with excessive memory usage.

The bug is reproducible if running deno from `C:\`, `C:\Windows`, `C:\Users`, `C:\Users\<current_user>`, `C:\Windows\System32`, `C:\Windows\SysWow64`. At the same time runinng `deno repl` from `C:\Users\<current_user>\Desktop` or `C:\Windows\System32\wbem` proceeds as normal. These are the directories that I've tested which reproduced the bug 100% of the times. Other directories and drives seems unaffected.

The last unaffected canary version with available binaries to download is `deno 1.32.1+e888c3f`

Edit:
Just checked with the 1.32.2 release. It's still affected.

Edit2:
Checked 1.32.2 Linux under Windows WSL, it hangs when executing with the root dir (`/`) as the current directory.

### PR DESCRIPTION
**Pull request by dsherret**
I was testing this out and it's badly needed. For now, it's not configurable and limited to 1,000 file system entries.

Related to #18538

### CODE REVIEW

File: cli/lsp/documents.rs [hunk 1]
```diff
@@ -1632,45 +1604,113 @@ impl PreloadDocumentFinder {
...
+      if let Some(dir_name) = dir_path.file_name() {
+        let dir_name = dir_name.to_string_lossy().to_lowercase();
+        // We ignore these directories by default because there is a
+        // high likelihood they aren't relevant. Someone can opt-into
+        // them by specifying one of them as an enabled path.
+        if matches!(dir_name.as_str(), "node_modules" | ".git") {
+          return false;
+        }
+
+        // ignore cargo target directories for anyone using Deno with Rust
```
Comment (dsherret) [inline cli/lsp/documents.rs:1630]: I think it's worth it to look for and handle these kind of scenarios.

File: cli/lsp/documents.rs [hunk 2]
```diff
@@ -1632,45 +1604,113 @@ impl PreloadDocumentFinder {
...
+
+            if self.entry_count >= self.limit {
+              lsp_warn!(
+                concat!(
+                  "Hit the language server document preload limit of {} file system entries. ",
+                  "You may want to use the \"deno.enablePaths\" configuration setting to only have Deno ",
+                  "partially enable a workspace."
+                ),
+                self.limit,
+              );
```
Comment (bartlomieju) [inline cli/lsp/documents.rs:1686]: Maybe in `vscode_deno` we could show a pop up if this message is printed?

**Review by bartlomieju — APPROVED** [review]
LGTM, thanks for fixing this

File: cli/lsp/documents.rs (hunk 2 above)
Comment (dsherret) [reply cli/lsp/documents.rs:1686]: Let's iterate on this in the future. I'd have to refactor the code to get the lsp client here and perhaps some users might not like this message.

**Pull request merged**